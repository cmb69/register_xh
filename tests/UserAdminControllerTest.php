<?php

/**
 * Copyright (c) 2023 Christoph M. Becker
 *
 * This file is part of Register_XH.
 */

namespace Register;

use ApprovalTests\Approvals;
use PHPUnit\Framework\TestCase;
use XH\CSRFProtection as CsrfProtector;

use Register\Value\User;
use Register\Value\UserGroup;
use Register\Infra\DbService;
use Register\Infra\Request;
use Register\Infra\Url;

class UserAdminControllerTest extends TestCase
{
    /** @var Request */
    private $request;

    public function setUp(): void
    {
        $this->request = $this->createStub(Request::class);
        $this->request->method("url")->willReturn(new Url("/", "Foo"));
    }

    public function testEditUsersActionRendersUsers()
    {
        $dbService = $this->createStub(DbService::class);
        $dbService->method('hasUsersFile')->willReturn(true);
        $dbService->method('readUsers')->willReturn([$this->makeJohn(), $this->makeJane()]);
        $dbService->method('hasGroupsFile')->willReturn(true);
        $dbService->method('readGroups')->willReturn([new UserGroup("users", ""), new UserGroup("guest", "")]);
        $sut = $this->makeUserAdminController($dbService);
        $response = $sut->editUsersAction($this->request);
        Approvals::verifyHtml($response->output());
    }

    public function testEditUsersActionIncludesScripts()
    {
        $dbService = $this->createStub(DbService::class);
        $dbService->method('hasUsersFile')->willReturn(true);
        $dbService->method('readUsers')->willReturn([$this->makeJohn(), $this->makeJane()]);
        $dbService->method('hasGroupsFile')->willReturn(true);
        $dbService->method('readGroups')->willReturn([new UserGroup("users", "")]);
        $sut = $this->makeUserAdminController($dbService);
        $response = $sut->editUsersAction($this->request);
        $expected = [
            "register_texts" => [
                "name" => "Full Name",
                "username" => "Username",
                "password" => "Password",
                "email" => "Email",
                "accessgroups" => "Access Groups",
                "status" => "Status",
                "prefsemailsubject" => "Account data changed for",
                "confirmLeave" => "Unsaved changes will be lost!",
                "newPassword" => "New Password (may contain only alphanumeric characters and underscores):",
                "tooManyUsers" => "There are too many users!",
            ],
            "register_max_number_of_users" => 142,
        ];
        $this->assertEquals($expected, $response->meta());
        $this->assertEquals("./admin.min.js", $response->script());
    }

    public function testEditUsersActionFailsIfNoUsersFile()
    {
        $dbService = $this->createStub(DbService::class);
        $dbService->method('hasUsersFile')->willReturn(false);
        $sut = $this->makeUserAdminController($dbService);
        $response = $sut->editUsersAction($this->request);
        Approvals::verifyHtml($response->output());
    }

    public function testSaveUsersCanAddRecord()
    {
        $_POST = [
            "add" => [""],
            "username" => ["cmb"],
            "password" => ["test"],
            "oldpassword" => ["test"],
            "name" => ["Christoph Becker"],
            "email" => ["chris@example.com"],
            "accessgroups" => ["users"],
            "status" => ["activated"],
        ];
        $dbService = $this->createStub(DbService::class);
        $dbService->expects($this->never())->method('writeUsers');
        $dbService->method('hasGroupsFile')->willReturn(true);
        $dbService->method('readGroups')->willReturn([new UserGroup("users", "")]);
        $sut = $this->makeUserAdminController($dbService);
        $response = $sut->saveUsersAction($this->request);
        Approvals::verifyHtml($response->output());
    }

    public function testSaveUsersCanDeleteRecord()
    {
        $_POST = [
            "delete" => [0 => "1"],
            "username" => ["to_be_deleted", "cmb"],
            "password" => ["to_be_deleted", "test"],
            "oldpassword" => ["to_be_deleted", "test"],
            "name" => ["Toby Deleted", "Christoph Becker"],
            "email" => ["to_be_deleted@example.com", "chris@example.com"],
            "accessgroups" => ["", "users"],
            "status" => ["activated", "activated"],
        ];
        $dbService = $this->createStub(DbService::class);
        $dbService->expects($this->never())->method('writeUsers');
        $dbService->method('hasGroupsFile')->willReturn(true);
        $dbService->method('readGroups')->willReturn([new UserGroup("users", "")]);
        $sut = $this->makeUserAdminController($dbService);
        $response = $sut->saveUsersAction($this->request);
        Approvals::verifyHtml($response->output());
    }

    public function testSaveUsersFailsOnInvalidUserName()
    {
        $_POST = [
            "username" => [""],
            "password" => ["test"],
            "oldpassword" => ["test"],
            "name" => ["Christoph Becker"],
            "email" => ["cmb@example.com"],
            "accessgroups" => ["users"],
            "status" => ["activated"],
        ];
        $dbService = $this->createStub(DbService::class);
        $dbService->expects($this->never())->method('writeUsers');
        $dbService->method('hasGroupsFile')->willReturn(true);
        $dbService->method('readGroups')->willReturn([new UserGroup("users", "")]);
        $sut = $this->makeUserAdminController($dbService);
        $response = $sut->saveUsersAction($this->request);
        Approvals::verifyHtml($response->output());
    }

    public function testSaveUsersFailsOnInvalidUserNameWhenChangingPassword()
    {
        $_POST = [
            "username" => [""],
            "password" => ["new"],
            "oldpassword" => ["old"],
            "name" => ["Christoph Becker"],
            "email" => ["cmb@example.com"],
            "accessgroups" => ["users"],
            "status" => ["activated"],
        ];
        $dbService = $this->createStub(DbService::class);
        $dbService->expects($this->never())->method('writeUsers');
        $dbService->method('hasGroupsFile')->willReturn(true);
        $dbService->method('readGroups')->willReturn([new UserGroup("users", "")]);
        $sut = $this->makeUserAdminController($dbService);
        $response = $sut->saveUsersAction($this->request);
        Approvals::verifyHtml($response->output());
    }

    public function testSaveUsersFailsOnDuplicateUserNameAndEmail()
    {
        $_POST = [
            "username" => ["cmb", "cmb"],
            "password" => ["pw1", "pw2"],
            "oldpassword" => ["pw1", "pw2"],
            "name" => ["Christoph Becker", "Christoph Becker"],
            "email" => ["cmb@example.com", "cmb@example.com"],
            "accessgroups" => ["users", "users"],
            "status" => ["activated", "activated"],
        ];
        $dbService = $this->createStub(DbService::class);
        $dbService->expects($this->never())->method('writeUsers');
        $dbService->method('hasGroupsFile')->willReturn(true);
        $dbService->method('readGroups')->willReturn([new UserGroup("users", "")]);
        $sut = $this->makeUserAdminController($dbService);
        $response = $sut->saveUsersAction($this->request);
        Approvals::verifyHtml($response->output());
    }

    public function testSaveUsersSuccessfullySaves()
    {
        $_POST = [
            "username" => ["cmb"],
            "password" => ["pw"],
            "oldpassword" => ["pw"],
            "name" => ["Christoph Becker"],
            "email" => ["cmb@example.com"],
            "accessgroups" => ["users"],
            "status" => ["activated"],
        ];
        $dbService = $this->createStub(DbService::class);
        $dbService->expects($this->once())->method('writeUsers')->willReturn(true);
        $dbService->method('hasGroupsFile')->willReturn(true);
        $dbService->method('readGroups')->willReturn([new UserGroup("users", "")]);
        $sut = $this->makeUserAdminController($dbService);
        $response = $sut->saveUsersAction($this->request);
        Approvals::verifyHtml($response->output());
    }
    private function makeUserAdminController(DbService $dbService): UserAdminController
    {
        return new UserAdminController(
            "./",
            $this->makeConf(),
            $this->makeLang(),
            $this->makeCsrfProtector(false),
            $dbService,
            "/"
        );
    }

    /** @return array<string,string> */
    private function makeConf(): array
    {
        $plugin_cf = XH_includeVar("./config/config.php", 'plugin_cf');
        return $plugin_cf['register'];
    }

    /** @return array<string,string> */
    private function makeLang(): array
    {
        $plugin_tx = XH_includeVar("./languages/en.php", 'plugin_tx');
        return $plugin_tx['register'];
    }

    private function makeCsrfProtector(bool $used): CsrfProtector
    {
        $csrfProtector = $this->createStub(CsrfProtector::class);
        if ($used) {
            $csrfProtector->expects($this->once())->method('check');
        }
        return $csrfProtector;
    }

    private function makeJohn(): User
    {
        return new User(
            "john",
            '$2y$10$gOae/VL5wrESo5Uf6ZcWhuNlAEycCGW5Ov5opny5PWxa.gbl4SHQW',
            ["guest"],
            "John Doe",
            "john@example.com",
            "activated"
        );
    }

    private function makeJane(): User
    {
        return new User(
            "jane",
            '$2y$10$gOae/VL5wrESo5Uf6ZcWhuNlAEycCGW5Ov5opny5PWxa.gbl4SHQW',
            ["admin"],
            "Jane Doe",
            "jane@example.com",
            "activated"
        );
    }
}
