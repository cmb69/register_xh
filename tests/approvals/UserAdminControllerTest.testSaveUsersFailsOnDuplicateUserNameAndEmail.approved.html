
<p class="xh_fail">Error for user 'cmb':</p>
<p class="xh_fail">The chosen username exists already.</p>
<p class="xh_fail">A user with the given email address exists already.</p>
<h1>User Administration</h1>
<div class="register_admin_main">
  <table>
    <tr id="register_user_template" style="display: none">
      <td><button title="Delete User" onclick="register.removeRow(this); return false">Delete</button></td>
      <td><input type="text" value="" name="name[]"></td>
      <td><input type="text" value="guest" name="accessgroups[]"></td>
      <td><select name="status[]"><option value="activated" selected="selected">activated</option><option value="locked">locked</option><option value="">deactivated</option></select></td>
    </tr>
    <tr style="display: none">
      <td></td>
      <td><input type="text" value="" name="username[]"></td>
      <td><input type="email" value="" name="email[]"></td>
      <td>
        <button onclick="register.changePassword(this.nextElementSibling); return false">Change Password...</button>
        <input type="hidden" value="" name="password[]">
        <input type="hidden" value="" name="oldpassword[]">
      </td>
    </tr>
  </table>
  <div>
    <button onclick="register.addRow()">Add User</button>
    <input id="register_toggle_details" type="checkbox" onclick="register.toggleDetails()" style="padding-left: 1em">
    <label for="register_toggle_details">Details</label>
    <select id="register_group_selectbox" title="Filter by access group">
      <option value="">All</option>
      <option value="users">users</option>
    </select>
  </div>
  <form id="register_user_form" method="post" action="/?register">
    <input type="hidden" value="saveusers" name="action">
    <input type="hidden" value="plugin_main" name="admin">
        <table id="register_user_table">
      <tr>
        <th></th>
        <th class="register_sort" onclick="register.sort(this, 'name')" style="cursor: pointer">Full Name</th>
        <th class="register_sort" onclick="register.sort(this, 'accessgroups')" style="cursor: pointer">Access Groups</th>
        <th class="register_sort" onclick="register.sort(this, 'status')" style="cursor: pointer">Status</th>
      </tr>
      <tr class="register_second_row">
        <th></th>
        <th class="register_sort" onclick="register.sort(this, 'username')" style="cursor: pointer">Username</th>
        <th class="register_sort" onclick="register.sort(this, 'email')" style="cursor: pointer">Email</th>
        <th>Password</th>
      </tr>
      <tr id="register_user_0">
        <td><button title="Delete User" onclick="register.removeRow(this); return false">Delete</button></td>
        <td><input type="text" value="Christoph Becker" name="name[0]"></td>
        <td><input type="text" value="users" name="accessgroups[0]"></td>
        <td><select name="status[0]"><option value="activated" selected="selected">activated</option><option value="locked">locked</option><option value="">deactivated</option></select></td>
      </tr>
      <tr class="register_second_row">
        <td><button type="button" onclick="register.mailTo(this)" title="Email">Mail</i></button></td>
        <td><input type="text" value="cmb" name="username[0]"></td>
        <td><input type="email" value="cmb@example.com" name="email[0]"></td>
        <td>
          <button onclick="register.changePassword(this.nextElementSibling); return false">Change Password...</button>
          <input type="hidden" value="pw1" name="password[0]">
          <input type="hidden" value="pw1" name="oldpassword[0]">
        </td>
      </tr>
      <tr id="register_user_1">
        <td><button title="Delete User" onclick="register.removeRow(this); return false">Delete</button></td>
        <td><input type="text" value="Christoph Becker" name="name[1]"></td>
        <td><input type="text" value="users" name="accessgroups[1]"></td>
        <td><select name="status[1]"><option value="activated" selected="selected">activated</option><option value="locked">locked</option><option value="">deactivated</option></select></td>
      </tr>
      <tr class="register_second_row">
        <td><button type="button" onclick="register.mailTo(this)" title="Email">Mail</i></button></td>
        <td><input type="text" value="cmb" name="username[1]"></td>
        <td><input type="email" value="cmb@example.com" name="email[1]"></td>
        <td>
          <button onclick="register.changePassword(this.nextElementSibling); return false">Change Password...</button>
          <input type="hidden" value="pw2" name="password[1]">
          <input type="hidden" value="pw2" name="oldpassword[1]">
        </td>
      </tr>
    </table>
    <input class="submit" type="submit" value="Save" name="send">
  </form>
</div>
<script type="text/javascript">register.init()</script>
